### PR TITLE
Further merge query types into entity types

### LINKED ISSUE #14195: ModelBuilder: QueryTypeBuilder.ToView continuously generates EnsureSchema in migrations.
When binding a `QueryType` to a view or table using `QueryTypeBuilder.ToView` and specifying a schema, an `EnsureSchema` statement is created for the `QueryType` in every subsequent migration. It doesn't break anything, but it does get a little bit tedious to keep picking the rogue `EnsureSchemas` out of migration scripts.

### Steps to reproduce
1. Create a DbContext with a `QueryType`.

```c#
public class SomeType
{
    public string Column1 { get; set; }
}

public class SampleContext : DbContext
{
    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.Query<SomeType>(builder =>
        {
            builder.ToView(schema: "SomeSchema", name: "Things");
        });
    }
}
```
2. Run `dotnet ef migrations add Test`
    - I'm expecting the model snapshot to have some reference to `EnsureSchema` or to maybe even just have an empty migration since ensuring that the schema exists doesn't really help if the query is relying on existence of some SQL table or view that isn't being auto-generated by the migration
3. Run `dotnet ef migrations add TestAgain`
     - Both migrations contain an `EnsureSchema` statement
     ```c#
     protected override void Up(MigrationBuilder migrationBuilder)
     {
         migrationBuilder.EnsureSchema(
             name: "SomeSchema");
     }
     ```

[EfCoreViewWithSchema.zip](https://github.com/aspnet/EntityFrameworkCore/files/2687883/EfCoreViewWithSchema.zip)

### Further technical details
EF Core version: 2.2.0
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 10
IDE: Visual Studio 2019 16.0.0 Preview 1.1

### LINKED ISSUE #14787: EntityTypeBuilder ToView or not ToView
Should it remain obsoleted or become a permanent synonym for `ToTable`?

### PR DESCRIPTION
**Pull request by bricelam**
Changes:
- Un-obsolete ToView
- Differentiate between views and keyless tables in RevEng
- Ignore ToView and ToQuery in Migrations; handle HasNoKey
- Handle duplicate table names when HasNoKey

Resolves #14787, fixes #14195

### CODE REVIEW

**Comment by bricelam** [comment]
Stubbing out parts of #14537 as an implementation detail.

File: src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs [hunk 1]
```diff
@@ -520,15 +522,17 @@ private static string EscapeLiteral(string s)
...
 
-                        var table = new DatabaseTable
-                        {
-                            Schema = schema,
-                            Name = name,
-                            Comment = comment,
-                        };
+                        var table = type == "table"
+                            ? new DatabaseTable()
+                            : new DatabaseView();
```
Comment (bricelam) [inline src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs]: cc @ErikEJ

Comment (ErikEJ) [reply src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs]: LGTM 😎

File: test/EFCore.Design.Tests/Migrations/ModelSnapshotSqlServerTest.cs [hunk 2]
```diff
@@ -271,6 +271,7 @@ public virtual void Entities_are_stored_in_model_snapshot()
                 },
                 AddBoilerPlate(
                     GetHeading() + @"
+
```
Comment (AndriySvyryd) [inline test/EFCore.Design.Tests/Migrations/ModelSnapshotSqlServerTest.cs]: Add the extra line to Heading?

**Comment by AndriySvyryd** [comment]
As per https://github.com/aspnet/EntityFrameworkCore/pull/17044#pullrequestreview-273242542 We should consider changing the condition in `TryUniquifyTableNames` to only skip entities mapped to views.

File: test/EFCore.Design.Tests/Migrations/ModelSnapshotSqlServerTest.cs (hunk 2 above)
Comment (bricelam) [reply test/EFCore.Design.Tests/Migrations/ModelSnapshotSqlServerTest.cs]: Needed it to not be there for one test ☹️

Comment (AndriySvyryd) [reply test/EFCore.Design.Tests/Migrations/ModelSnapshotSqlServerTest.cs]: You could add a bool parameter to `GetHeading()`

**Comment by bricelam** [comment]
@AndriySvyryd fixed that, and the duplicate table name validation

**Comment by AndriySvyryd** [comment]
:shipit:

File: test/EFCore.Relational.Specification.Tests/Query/InheritanceRelationalFixture.cs [hunk 3]
```diff
@@ -30,7 +30,8 @@ protected override void OnModelCreating(ModelBuilder modelBuilder, DbContext con
             modelBuilder.Entity<Lilt>().Property(e => e.SugarGrams).HasColumnName("SugarGrams");
             modelBuilder.Entity<Tea>().Property(e => e.CaffeineGrams).HasColumnName("CaffeineGrams");
 
-            modelBuilder.Entity<AnimalQuery>().HasNoKey().ToTable("Animal");
+            modelBuilder.Entity<AnimalQuery>().HasNoKey().ToQuery(
+                () => context.Set<AnimalQuery>().FromSqlRaw("SELECT * FROM Animal"));
```
Comment (bricelam) [inline test/EFCore.Relational.Specification.Tests/Query/InheritanceRelationalFixture.cs]: @maumar FYI, I had to switch some tests to this pattern and skip them until #16323 is implemented.

Comment (bricelam) [reply test/EFCore.Relational.Specification.Tests/Query/InheritanceRelationalFixture.cs]: (We now include keyless entity types when validating that two unrelated entity types don't map to the same table.)

**Pull request merged**